lib/auth/auth_with_roles_test.go

```diff
@@ -3943,114 +3943,92 @@ func TestGenerateHostCert(t *testing.T) {
 	}
 }
 
-// TestLocalServiceRolesHavePermissionsForUploaderService verifies that all of Teleport's
-// builtin roles have permissions to execute the calls required by the uploader service.
-// This is because only one uploader service runs per Teleport process, and it will use
-// the first available identity.
-func TestLocalServiceRolesHavePermissionsForUploaderService(t *testing.T) {
+// TestInstanceRoleCanUpload verifies that Teleport's Instance role
+// has permissions to execute the calls required by the uploader service.
+func TestInstanceRoleCanUpload(t *testing.T) {
 	srv, err := NewTestAuthServer(TestAuthServerConfig{Dir: t.TempDir()})
 	require.NoError(t, err)
 
-	// Test all local service roles, plus RoleInstance.
-	// The latter may also be used to run the uploader.
-	roles := append(types.LocalServiceMappings(), types.RoleInstance)
-	for _, role := range roles {
-		// RoleMDM services don't create events by themselves, instead they rely on
-		// Auth to issue events.
-		if role == types.RoleAuth || role == types.RoleMDM {
-			continue
-		}
-		t.Run(role.String(), func(t *testing.T) {
-			ctx := context.Background()
-
-			var identity TestIdentity
-			if role == types.RoleInstance {
-				// RoleInstance needs AdditionalSystemRoles, otherwise the setup is the
-				// same.
-				identity = TestIdentity{
-					I: authz.BuiltinRole{
-						Role: role,
-						AdditionalSystemRoles: []types.SystemRole{
-							types.RoleNode, // Arbitrary, could be any role.
-						},
-						Username: string(role),
-					},
-				}
-			} else {
-				identity = TestBuiltin(role)
-			}
+	ctx := context.Background()
+	identity := TestIdentity{
+		I: authz.BuiltinRole{
+			Role: types.RoleInstance,
+			AdditionalSystemRoles: []types.SystemRole{
+				types.RoleNode, // Arbitrary, could be any role.
+			},
+			Username: string(types.RoleInstance),
+		},
+	}
```

Comment:
nit: effectively this test is verifying that the node role has the permission to upload (and that instance certs aren't bugged).  Instance isn't a role in the sense of `RoleAuth` or `RoleNode`, it's a placeholder used to indicate that a cert contains multiple system roles.  An instance cert has the combined permissions of all true system roles that are listed on the cert.

A quick/easy way to make this much more powerful would be to make this test a sub-test that takes the value of the "additional system role" as a placeholder.  Then the outer test could iterate and run this for all the system roles that we want to assert can upload.

Reply:
Ok, reverted back to how it used to be (table test over all local service mapping roles) only instead of using the role directly we use the role "inside" an instance role.